nmagent/cachedclient_test.go

```diff
@@ -0,0 +1,87 @@
+package nmagent_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/Azure/azure-container-networking/nmagent"
+	"github.com/Azure/azure-container-networking/nmagent/fakes"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+)
+
+// TestHomeAzCache makes sure the CachedClient works properly in caching home ez response and error
+func TestHomeAzCache(t *testing.T) {
+	tests := []struct {
+		name         string
+		client       *fakes.NMAgentClientFake
+		getHomeAzExp nmagent.HomeAzResponse
+		shouldErr    bool
+	}{
+		{
+			"happy path",
+			&fakes.NMAgentClientFake{
+				SupportedAPIsF: func(ctx context.Context) ([]string, error) {
+					return []string{"GetHomeAz"}, nil
+				},
+				GetHomeAzF: func(ctx context.Context) (nmagent.HomeAzResponse, error) {
+					return nmagent.HomeAzResponse{IsSupported: true, HomeAz: uint(1)}, nil
+				},
+			},
+			nmagent.HomeAzResponse{IsSupported: true, HomeAz: uint(1)},
+			false,
+		},
+		{
+			"getHomeAz is not supported in nmagent",
+			&fakes.NMAgentClientFake{
+				SupportedAPIsF: func(ctx context.Context) ([]string, error) {
+					return []string{"dummy"}, nil
+				},
+				GetHomeAzF: func(ctx context.Context) (nmagent.HomeAzResponse, error) {
+					return nmagent.HomeAzResponse{}, nil
+				},
+			},
+			nmagent.HomeAzResponse{},
+			false,
+		},
+		{
+			"unexpected errors",
+			&fakes.NMAgentClientFake{
+				SupportedAPIsF: func(ctx context.Context) ([]string, error) {
+					return []string{"dummy"}, nil
+				},
+				GetHomeAzF: func(ctx context.Context) (nmagent.HomeAzResponse, error) {
+					return nmagent.HomeAzResponse{}, errors.New("unexpected error")
+				},
+			},
+			nmagent.HomeAzResponse{},
+			false,
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			client := nmagent.NewCachedClient(test.client)
+			client.Start(1)
+
+			// give some time for the thread to complete retrieving home az and update the cache
+			time.Sleep(2 * time.Second)
```

Comment:
That this is necessary highlights a temporal dependency that shouldn't be here. In reality, a caller could invoke any method immediately after creating the client. Nothing is stopping them.

Also, `time.Sleep` in general is a smell in tests (no matter how much we do it elsewhere). It slows the tests down by default. Also, by failing to address the underlying ordering-of-events problem, it usually means that eventually something else will be added and require even more `time.Sleep` (repeat ad infinitum).